rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,649 @@
+
+  """A python variable from an existing handle."""
+  def _read_variable_op(self):
+
+    # The `assign_variable_op` will be called to get before read opertion.
+    # `prefetch_values` is a lookup-like operation of `TrainableWrap`.
+
+    if hasattr(self, "prefetch_values"):
+      with ops.control_dependencies([
+          gen_resource_variable_ops.assign_variable_op(
+            self._handle, self.prefetch_values,  
+          name="AssignBeforeReadVariable")]):
+        result = gen_resource_variable_ops.read_variable_op(self._handle,
+                                                            self._dtype)
+    else:
+      result = gen_resource_variable_ops.read_variable_op(self._handle,
+                                                          self._dtype)
+    _maybe_set_handle_data(self._dtype, self._handle, result)
+
+
+  def update_op(self):
+    return self.params.upsert(self.ids, self._read_variable_op())
+```
+
+The scheme will **reuse** the `Optimizer._resource_apply_{dense,sparse_duplicate_indices}}` API to train `resource_variable_ops.TrainableWrap` naturally after changing a little common code in `training/optimizer.py`:
+
+- Create the slots of `TrainableWrap` when primary `Variable` is an instance of `TrainableWrap`.
+- Write back the trained values of primary and slots to hash tables(`dynamic_embedding.Variable`) when `Optimizer._resource_apply_{dense,sparse_duplicate_indices}` finished.
+
+#### The pseudo-code is below: 
+
+```python
+@tf_export("dynamic_embedding.create_slots")
+def create_slots(primary,
+                 init,
+                 slot_name,
+                 op_name):
+  """Helper function for creating a slot variable for statefull optimizers."""
+  _params_var, _params_ids = _IDS_OF_RESOURCEVARIABLE_MAPPER_.get(primary)
+  with variable_scope.variable_scope(None, _params_var.name + "/" + op_name):
+    _slot_variable = Variable(key_dtype=_params_var.key_dtype,
+                              value_dtype=_params_var.value_dtype,
+                              dim=_params_var.dim,
+                              devices=_params_var.devices,
+                              partitioner=_params_var.partition_fn,
+                              shared_name="shared-" + slot_name,
+                              name=slot_name,
+                              initial_value=init,
+                              initial_mode="constant",
+                              trainable=False,
+                              checkpoint=_params_var.checkpoint)
+    slot = None
+    slot = embedding_lookup(params=_slot_variable,
+                            ids=_params_ids)
+    if _SLOT_MAP_.get(primary) is None:
+      _SLOT_MAP_[primary] = [slot]
+    else:
+      _SLOT_MAP_[primary].append(slot)
+
+  return slot
+
+
+@tf_export("dynamic_embedding.get_slots")
+def get_slots(primary):
+  return _SLOT_MAP_.get(primary) or []
+
+```
+
+
+```python
+
+class _DenseDynamicEmbeddingTrainableProcessor(_OptimizableVariable):
```

Comment:
Does this class need to be different for different optimizers?

Reply:
No need to be.